### PR TITLE
NIFI-8111: Add JSLTTransformJSON processor

### PR DESCRIPTION
**Pull request by mattyb149**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Provides an alternative to JoltTransformJSON using [JSLT](https://github.com/schibsted/jslt) as the transformation language. The JOLT DSL can have a steep learning curve, where JSLT is inspired by other popular tools such as `jq`, `XPath` and `XQuery`. The resulting NAR is < 2MB and only has Jackson as a dependency so I included it in the assembly.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [x] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+            } else {
+                jsonString = context.getProperty(PRETTY_PRINT).asBoolean() ? transformedJson.toPrettyString() : transformedJson.toString();
+            }
+        } catch (final Exception ex) {
+            logger.error("Unable to transform {} due to {}", new Object[]{original, ex.toString()}, ex);
+            session.transfer(original, REL_FAILURE);
+            return;
+        }
+
+        FlowFile transformed = session.write(original, out -> out.write(jsonString.getBytes(DEFAULT_CHARSET)));
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java:223]: You could use 
```suggestion
        FlowFile transformed = session.write(original, out -> out.write(jsonString.getBytes(StandardCharsets.UTF_8)));
```
here.

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+        if (original == null) {
+            return;
+        }
+
+        final ComponentLog logger = getLogger();
+        final StopWatch stopWatch = new StopWatch(true);
+
+        JsonNode firstJsonNode;
+        try (final InputStream in = session.read(original)) {
+            try {
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java:179]: Please, consider extracting this nested try block into a separate method.

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/test/java/org/apache/nifi/processors/jslt/TestJSLTTransformJSON.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    @Test
+    public void testBadInput() {
+        final TestRunner runner = TestRunners.newTestRunner(new JSLTTransformJSON());
+        final String invalidTransform = "invalid";
+        runner.setProperty(JSLTTransformJSON.JSLT_TRANSFORM, invalidTransform);
+        runner.assertNotValid();
+    }
+
+    @Test
+    public void testSimpleJSLT() throws IOException {
```
Comment (Lehel44) [inline nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/test/java/org/apache/nifi/processors/jslt/TestJSLTTransformJSON.java:60]: Please, consider using  one parameterized test here to reduce duplicated code.

**Comment by mattyb149** [comment]
I'd like to reconsider this as a PR on its own without the JSLTTransformRecord processor, that one is more complicated as we'd need a full implementation of Jackson's `JsonParser` and related classes for NiFi record-based parsing, which is a significant effort versus this quick win for JSLT parsing if the input is JSON. If accepted I will write a follow-up Jira to add JSLTTransformRecord

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java (hunk 2 above)
Comment (MikeThomsen) [reply nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java:179]: I second that.

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+            }
+        } catch (final Exception e) {
+            logger.error("Failed to transform {}; routing to failure", new Object[]{original}, e);
+            session.transfer(original, REL_FAILURE);
+            return;
+        }
+
+        final String jsonString;
+        try {
+            Expression jsltExpression = precompiledExpression.get();
```
Comment (MikeThomsen) [inline nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java:204]: This feels like a good use case for something modeled on `RecordPathCache`. I think Caffeine provides something that can be easily used for this.

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/test/java/org/apache/nifi/processors/jslt/TestJSLTTransformJSON.java [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+    final static Path JSON_INPUT = Paths.get("src/test/resources/input.json");
+
+    @Test
+    public void testInvalidJSLTTransform() throws IOException {
+        final TestRunner runner = TestRunners.newTestRunner(new JSLTTransformJSON());
+        // This avoids an error under test where the OnScheduled method doesn't evaluate Expression Language on the transform property (after it verifies no EL is present)
+        runner.setValidateExpressionUsage(false);
+        final String inputFlowFile = "I am not JSON";
+        final String transform = new String(Files.readAllBytes(Paths.get("src/test/resources/simpleTransform.json")));
```
Comment (MikeThomsen) [inline nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/test/java/org/apache/nifi/processors/jslt/TestJSLTTransformJSON.java:42]: Is `simpleTransform` an invalid `JSLT`? I wasn't sure when looking at it.

**Review by MikeThomsen — CHANGES_REQUESTED** [review]
Rebase against main, and I think this is pretty close to being done.

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/test/java/org/apache/nifi/processors/jslt/TestJSLTTransformJSON.java (hunk 5 above)
Comment (mattyb149) [reply nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/test/java/org/apache/nifi/processors/jslt/TestJSLTTransformJSON.java:42]: It's a legit transform, it's the input that's invalid. I switched the names of the tests, will correct that.

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+@EventDriven
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java:67]: Should this annotation be removed?

File: README.md [hunk 7]
```diff
@@ -190,9 +190,9 @@ To run:
 
         $ cd minifi-assembly
         $ ls -lhd target/minifi*
-        drwxr-xr-x  3 user  staff   102B Jul  6 13:07 minifi-1.14.0-SNAPSHOT-bin
-        -rw-r--r--  1 user  staff    39M Jul  6 13:07 minifi-1.14.0-SNAPSHOT-bin.tar.gz
-        -rw-r--r--  1 user  staff    39M Jul  6 13:07 minifi-1.14.0-SNAPSHOT-bin.zip
+        drwxr-xr-x  3 user  staff   102B Jul  6 13:07 minifi-1.16.0-SNAPSHOT-bin
+        -rw-r--r--  1 user  staff    39M Jul  6 13:07 minifi-1.16.0-SNAPSHOT-bin.tar.gz
+        -rw-r--r--  1 user  staff    39M Jul  6 13:07 minifi-1.16.0-SNAPSHOT-bin.zip
```
Comment (exceptionfactory) [inline README.md:195]: It looks like these changes are unrelated to this pull request.

File: nifi-nar-bundles/nifi-asn1-bundle/nifi-asn1-services/src/test/resources/ASN1.xml [hunk 8]
```diff
@@ -241,7 +241,7 @@
                     <bundle>
                         <artifact>nifi-record-serialization-services-nar</artifact>
                         <group>org.apache.nifi</group>
-                        <version>1.14.0-SNAPSHOT</version>
+                        <version>1.16.0-SNAPSHOT</version>
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-asn1-bundle/nifi-asn1-services/src/test/resources/ASN1.xml:244]: The changes in this file appear unrelated.

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+@SideEffectFree
+@SupportsBatching
+@Tags({"json", "jslt", "transform"})
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@WritesAttribute(attribute = "mime.type", description = "Always set to application/json")
+@CapabilityDescription("Applies a JSLT transformation to the flowfile JSON payload. A new FlowFile is created "
+        + "with transformed content and is routed to the 'success' relationship. If the JSON transform "
+        + "fails, the original FlowFile is routed to the 'failure' relationship.")
+public class JSLTTransformJSON extends AbstractProcessor {
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java:76]: It looks like this processor reads the entire FlowFile content into memory when parsing JSON, similar to JoltTransformJSON.  For this reason, it would be helpful to add the memory resource consideration annotation.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
This looks like a useful new component @mattyb149. Aside from the existing comments, I noted a few additional recommendation.  I also noticed the test failures on Windows, which may be related to line-endings.  It would be good to adjust the tests or expected output so that the tests pass on all platforms. The underlying library has not been updated for over a year, but with transitive dependencies limited to Jackson, that does not seem to be a significant concern.

I also noted the potential memory concern related to reading the entire FlowFile into memory.  Although the JoltTransformJSON processor does the same thing, this is definitely a potential problem, so adding the resource consideration annotation would be helpful.

**Comment by mattyb149** [comment]
Had to force-push this because of the irrelevant file changes plus I rebased against the latest main. Hopefully this is good to go after a final review @Lehel44 @MikeThomsen @exceptionfactory  Thanks!

File: nifi-assembly/pom.xml [hunk 10]
```diff
@@ -862,6 +862,12 @@ language governing permissions and limitations under the License. -->
             <version>1.18.0-SNAPSHOT</version>
             <type>nar</type>
         </dependency>
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-jslt-nar</artifactId>
+            <version>1.16.0-SNAPSHOT</version>
```
Comment (exceptionfactory) [inline nifi-assembly/pom.xml:868]: Looks like this needs to be updated to `1.18.0-SNAPSHOT`.

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java (hunk 6 above)
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java:69]: The EventDriven annotation should be removed.

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+        }
+
+        final ComponentLog logger = getLogger();
+        final StopWatch stopWatch = new StopWatch(true);
+
+        JsonNode firstJsonNode;
+        try (final InputStream in = session.read(original)) {
+            firstJsonNode = readJson(in);
+        } catch (final Exception e) {
+            logger.error("Failed to transform {}; routing to failure", new Object[]{original}, e);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java:203]: The wrapping Object[] should be removed.

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+        final String jsonString;
+        final String transform;
+        try {
+            transform = context.getProperty(JSLT_TRANSFORM).evaluateAttributeExpressions(original).getValue();
+            Expression jsltExpression = transformCache.get(transform, currString -> Parser.compileString(transform));
+
+            final JsonNode transformedJson = jsltExpression.apply(firstJsonNode);
+            if (transformedJson == null) {
+                jsonString = "";
+                logger.info("JSLT transform resulted in no data!");
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java:217]: Log message should not include an exclamation mark. It is probably better to remove this line. Should this throw an exception, or at least log a warning?

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for revisiting this PR @mattyb149. I noted a couple items that need to be addressed, and it looks like it needs to be rebased again. Some of the implementation needs a little closer look, but this seems like a very useful new Processor.

File: nifi-assembly/pom.xml (hunk 10 above)
Comment (mattyb149) [reply nifi-assembly/pom.xml:868]: Will update to 1.19.0-SNAPSHOT

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+     *
+     * @param in the InputStream from which to read the input
+     * @return the top-level JsonNode for the parsed JSON object
+     * @throws IOException if an error occurs while parsing JSON from the InputStream
+     */
+    private JsonNode readJson(final InputStream in) throws IOException {
+        JsonNode firstJsonNode;
+        try {
+            JsonParser jsonParser = jsonFactory.createParser(in);
+            jsonParser.setCodec(codec);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java:251]: Instead of calling `JsonFactory.createParser()`, it looks like `ObjectMapper.readTree(InputStream)` should return the root JSON Node.

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+
+            JsonToken token = jsonParser.nextToken();
+            if (token == JsonToken.START_ARRAY) {
+                token = jsonParser.nextToken(); // advance to START_OBJECT token
+            }
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java:256]: This approach only reads the first element in an array. It seems like this could be confusing behavior, should the Processor require all incoming JSON to be an Object at the root level? If JSLT does not support transforming array root elements, then it seems better to enforce a standard. It may be worth noting this detail in the description one way or the other.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the updates @mattyb149, this looks close to completion. On further review, I noted a couple dependencies that seem unnecessary, and I raised a question about the input type of JSON elements.

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java (hunk 14 above)
Comment (mattyb149) [reply nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java:256]: Great catch! I want it to support array root elements like JoltTransformJSON does. One difference between JoltTransformJSON and JoltTransformRecord is that for the former you write the transform for the entire input (so if you want to iterate over records you need to put that `*` in the spec, and for the latter you write the transform for each object (record). I want to have the same behavior for JSLTTransformJSON, and someday would like to add a JSLTTransformRecord in the same vein. Will update this (and the above comment) to apply the transform to the entire input and add a unit test with a top-level array.

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/pom.xml [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-utils</artifactId>
+            <version>1.19.0-SNAPSHOT</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-record</artifactId>
+            <version>1.19.0-SNAPSHOT</version>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (mattyb149) [reply nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/pom.xml:42]: No, it is a remnant from when I tried to add this and a JSLTTransformRecord processor, good catch!

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+                outputObject = transformedJson;
+            }
+            FlowFile transformed = session.write(original, out -> {
+                if (outputObject != null) {
+                    writer.writeValue(out, outputObject);
+                }
+            });
+            transformed = session.putAttribute(transformed, CoreAttributes.MIME_TYPE.key(), "application/json");
+            session.transfer(transformed, REL_SUCCESS);
+            session.getProvenanceReporter().modifyContent(transformed, "Modified With " + transform, stopWatch.getElapsed(TimeUnit.MILLISECONDS));
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java:223]: Is this manual call to Provenance Reporter still necessary?

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the updates @mattyb149, the only remaining items is the call to Provenance Reporter, which seems like it could be removed, as it should be handled automatically.

If you can double-check that, I will plan on moving forward with merging.

File: nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java (hunk 16 above)
Comment (mattyb149) [reply nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java:223]: There are a number of processors that do their own modifyContent in order to specify a particular message with more context. This one I copied from JoltTransformJSON, so I kept it for consistency.

Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-jslt-bundle/nifi-jslt-processors/src/main/java/org/apache/nifi/processors/jslt/JSLTTransformJSON.java:223]: Thanks for the explanation @mattyb149!

**Review by exceptionfactory — APPROVED** [review]
Thanks for working through the feedback and bringing this to completion @mattyb149! The latest version looks good! +1 merging
